File: 5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() < 1) {// Trả về chuỗi rỗng nếu s là null hoặc độ dài nhỏ hơn 1
            return ""; // Trả về chuỗi rỗng
        } // Khởi tạo biến để theo dõi vị trí và độ dài của chuỗi đối xứng dài nhất
        int start = 0, end = 0; // Vị trí bắt đầu và kết thúc của chuỗi đối xứng dài nhất
        for (int i = 0; i < s.length(); i++) { // Duyệt qua từng ký tự trong chuỗi
            int len1 = expandAroundCenter(s, i, i); // Kiểm tra đối xứng với một ký tự trung tâm
            int len2 = expandAroundCenter(s, i, i + 1); // Kiểm tra đối xứng với hai ký tự trung tâm
            int len = Math.max(len1, len2); // Lấy độ dài lớn nhất giữa hai trường hợp
            if (len > end - start) { // Nếu độ dài lớn hơn độ dài hiện tại
                start = i - (len - 1) / 2; // Cập nhật vị trí bắt đầu
                end = i + len / 2; // Cập nhật vị trí kết thúc
            }
        }
        return s.substring(start, end + 1);// Trả về chuỗi con từ vị trí bắt đầu đến kết thúc
    }

    private int expandAroundCenter(String s, int left, int right) {// Hàm mở rộng từ trung tâm để tìm chuỗi đối xứng
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {// Kiểm tra điều kiện biên và đối xứng
            left--;// Giảm left để mở rộng về phía bên trái
            right++;// Mở rộng về hai phía
        }
        return right - left - 1;// Trả về độ dài của chuỗi đối xứng tìm được
    }
}
// @lc code=end

File: 5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if not s:
            return ""

        start, end = 0, 0 # Biến start và end để lưu trữ vị trí bắt đầu và kết thúc của chuỗi con đối xứng dài nhất

        for i in range(len(s)): # Duyệt qua từng ký tự trong chuỗi
            len1 = self.expandAroundCenter(s, i, i) # Tính độ dài của chuỗi con đối xứng với i là tâm
            len2 = self.expandAroundCenter(s, i, i + 1) # Tính độ dài của chuỗi con đối xứng với i và i + 1 là tâm
            max_len = max(len1, len2) # Lấy độ dài lớn nhất giữa hai chuỗi con đối xứng
            if max_len > end - start:   # Nếu độ dài lớn nhất lớn hơn độ dài hiện tại
                start = i - (max_len - 1) // 2 # Cập nhật vị trí bắt đầu
                end = i + max_len // 2 

        return s[start:end + 1] # Trả về chuỗi con đối xứng dài nhất

    def expandAroundCenter(self, s, left, right): # Mở rộng từ tâm
        while left >= 0 and right < len(s) and s[left] == s[right]: # Kiểm tra xem hai ký tự ở hai bên có bằng nhau không
            left -= 1 # Giảm left để mở rộng về bên trái
            right += 1 # Tăng right để mở rộng về bên phải
        return right - left - 1 # Trả về độ dài của chuỗi con đối xứng
    # @lc code=end
